---
title: "generate_PGS_stats"
author: "Florian Schnitzer"
date: "`r Sys.Date()`"
output: html_document

params:
  r2_ref_data_folder: //wsl.localhost/Ubuntu-22.04/home/flo/projects/rsq-browser/local_files/data_local/
  output_path: //wsl.localhost/Ubuntu-22.04/home/flo/projects/rsq-browser/local_files/data_local/
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

### load libraries
library(Rsamtools)
library(data.table)
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(plotly)
#library(DT)

```


```{r load_data, include = FALSE}

###### Functions #######

## function to generate df with compressed information and correct format for the "overview plot"
get_overview_plot_df <- function(population_df) {
  
  ## create MAF column
  population_df$MAF <- ifelse(population_df$AF <= 0.5, population_df$AF, 1-population_df$AF)
  
  ## create bins based on MAF (detailed -> better for plotting)
  population_df <- population_df %>%
  select(-CHR, -POS, -REF, -ALT, -AF) %>%
  mutate(bin_detailed = cut(MAF, breaks = c(0.0000, 0.0005, 0.001, 0.002, 0.005, 0.010, 0.015,
                                     0.020, 0.035, 0.050, 0.100, 0.200, 0.300, 0.400, 0.500),
                    labels = c(0.0005, 0.001, 0.002, 0.005, 0.01, 0.015, 0.02,
                               0.035, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)))
  
  ## create vector for array_ref_panel identifiers
  identifier_vec <- colnames(population_df)
  identifier_vec <- identifier_vec[!str_detect(identifier_vec, pattern = c("_in"))]
  identifier_vec <- identifier_vec[!str_detect(identifier_vec, pattern = c("bin_"))]
  identifier_vec <- identifier_vec[!str_detect(identifier_vec, pattern = c("MAF"))]
  
  ## initiate df
  plot_df <- data.frame()
  
  ## iterate over unique combis and collect data in the desired format for plotting
  for (i in 1:length(identifier_vec)) {
    
    ## make identifier_vec elements accessible within dplyr functions
    id_for_dplyr <- sym(identifier_vec[i])
    
    ## select only data from one combi and remove NAs
    temp_stats <- population_df %>%
      select(!! id_for_dplyr, bin_detailed) %>%
      na.omit()
    
    ## then group by bin and calculate the mean for this combi per bin, add identifier as additional column
    temp_stats <- temp_stats %>%
      group_by(bin_detailed) %>%
      summarize(mean_r2 = mean(!! id_for_dplyr)) %>%
      mutate(identifier = identifier_vec[i])
    
    ## append to plot_df
    plot_df <- rbind(plot_df, temp_stats)
  }
  
  
  plot_df <- plot_df %>%
    ## change bin column to numeric
    mutate(bin_detailed = as.numeric(sub("0+$", "", as.character(bin_detailed)))) %>%
           ## split identifier to get columns with array and ref_panel names
           separate_wider_delim(cols = identifier, ## column to separate/split
                         delim = "_", ## delimiter to separate/split on
                         names = c("genotyping_array", "ref_panel"),  ## names of output columns from split results
                         cols_remove = FALSE) ## explicitly keep the input column
  
  return(plot_df)

}




#### MAIN ######

### get list of r2_ref data filepaths in the r2_ref_data_folder (params argument) with correct suffix + return full name (absolute path)
pop_files <- list.files(params$r2_ref_data_folder, pattern = "bi.snv.tab.gz$", full.names = T)

### get corresponding list of filenames
pop_filenames <- list.files(params$r2_ref_data_folder, pattern = "bi.snv.tab.gz$")


## for each population file generate plot_df and save to csv
for (i in 1:length(pop_files)) {
  
  pop_filename <- pop_filenames[i]

  pop_data <- fread(pop_files[i])
  
  ## CAVE: one small error in row 5809321 (not counting the header row) -> "0.65.343" in column "IO_1kg" in the biome.bi.snv.tab.gz
  ## remedy by setting this column as.numeric -> erroneous value will be replaced with NA (will cause a warning for this file)
  pop_data <- pop_data %>%
    mutate(IO_1kg = as.numeric(IO_1kg))
  
  plot_pop_dat <- get_overview_plot_df(pop_data)
  
  ### important: add extension to pop_filename -> so that it can be accessed in index with population_file parameter but does NOT overwrite this file
  outfile_name <- paste(pop_filename, ".plot.csv", sep = "")
  
  outfile_path <- paste(params$output_path, outfile_name, sep = "")
  
  
  write.table(plot_pop_dat, file = outfile_path, quote = F, row.names = F, sep = "\t")
  
}




#### test if figure can be created from plot_eur_df
# lineplot <- ggplot(plot_df_eur, aes(x=bin_detailed, y=mean_r2, color = identifier)) +
#   ### draw line plot
#   geom_line()+
#   ### highlight actual data points
#   geom_point(shape=0)+
#   ## use log-scale on x-axis and display x-axis labels without trailing zeros
#   scale_x_log10(labels = function(x) as.character(x))+
#   ## manually set y-axis ticks and set limits of the y-axis (include whole range from 0 to 1)
#   scale_y_continuous(breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1), limits = c(0,1))+
#   xlab("Minor Allele Frequency (MAF)")+
#   ylab("Average Imputation Quality")+
#   theme_minimal()
# 
# ### convert to plotly object
# lineplot_plotly <- ggplotly(lineplot)



#### Open questions for this plot #######

### how to add interactivity -> reference panel, genotyping array + download?
### how to include variant number (required?)
### how to map correct label (e.g. Core (1000 Genomes) [xxx variants])




```


```{r show_table, echo=FALSE}
lineplot_plotly
```









